OrbitCore/CallstackData.cpp

```diff
@@ -32,5 +117,51 @@ void CallstackData::AddCallStackFromKnownCallstackData(
     absl::MutexLock lock(&unique_callstacks_mutex_);
     unique_callstacks_[hash] = std::move(unique_callstack);
   }
-  callstack_events_.push_back(CallstackEvent(event));
+  absl::MutexLock lock(&callstack_events_by_tid_mutex_);
+  callstack_events_by_tid_[event.thread_id()][event.time()] = CallstackEvent(event);
+}
+
+const CallStack* CallstackData::GetCallStack(CallstackID callstack_id) const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  auto it = unique_callstacks_.find(callstack_id);
+  if (it != unique_callstacks_.end()) {
+    return it->second.get();
+  }
+  return nullptr;
+}
+
+bool CallstackData::HasCallStack(CallstackID callstack_id) const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  return unique_callstacks_.count(callstack_id) > 0;
+}
+
+void CallstackData::ForEachUniqueCallstack(
+    const std::function<void(const CallStack&)>& action) const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  for (const auto& it : unique_callstacks_) {
+    action(*it.second);
+  }
+}
+
+void CallstackData::ForEachFrameInCallstack(uint64_t callstack_id,
+                                            const std::function<void(uint64_t)>& action) const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  for (uint64_t frame : unique_callstacks_.at(callstack_id)->GetFrames()) {
+    action(frame);
+  }
+}
+
+absl::flat_hash_map<CallstackID, std::shared_ptr<CallStack>>
+CallstackData::GetUniqueCallstacksCopy() const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  return unique_callstacks_;
+}
+
+std::shared_ptr<CallStack> CallstackData::GetCallstackPtr(CallstackID callstack_id) const {
```

Comment:
Can this be const? I think we should make it clear, that you can not modify while capturing.

Reply:
Basically, I am aiming with CaptureData towards write only on the capture thread.

Reply:
What part should be const? This is a private method by the way, used only by `CallstackData::AddCallStackFromKnownCallstackData`, which is used when making a selection.

Reply:
I mean if it can return a const shared_ptr, such that you can only call const methods on the underlying Callstack

Reply:
Making the `shared_ptr` `const` doesn't achieve that, the pointer becomes const, not the pointee. I can try to make the type argument const but that will require adding const to many other places, let me see.

Reply:
If I'm not mistaken, CallStack is always const: you can modify it after initialization.

Reply:
True, it's immutable.